File: Binary_Tree_Maximum_Path_Sum.cpp
/**
 * Definition of TreeNode:
 * class TreeNode {
 * public:
 *     int val;
 *     TreeNode *left, *right;
 *     TreeNode(int val) {
 *         this->val = val;
 *         this->left = this->right = NULL;
 *     }
 * }
 */
class Solution {
public:
    /**
     * @param root: The root of binary tree.
     * @return: An integer
     */
    int maxPathSum(TreeNode *root) {
        // write your code here
        maxPathSumRecu(root);
        return max_sum_;
    }
    int maxPathSumRecu(TreeNode *root){
        if (!root) return 0;
        int left = max(0, maxPathSumRecu(root->left));
        int right = max(0, maxPathSumRecu(root->right));
        max_sum_ = max(max_sum_, root->val + left + right);
        return root->val + max(left, right);
    }
private:
    int max_sum_ = INT_MIN;
};

File: Binary Tree Maximum Path Sum.py
"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""
class Solution:
    """
    @param root: The root of binary tree.
    @return: An integer
    """
    def maxPathSum(self, root):
        # write your code here
        # def helper(root):
        #     if not root:
        #         return float('-inf'), float('-inf')
        #     else:
        #         (left, left_max), (right, right_max) = map(lambda x: helper(x), [root.left, root.right])
        #         leaf = max(left, right)
        #         cur = max(leaf + root.val, root.val)
        #         return (cur, max(cur, left_max, right_max, root.val + left + right))
        
        # __, ret = helper(root)
        # return ret
        
        def helper(root):
            if not root:
                ## error: 0 used to add, float('-inf') used to be compared
                return 0, float('-inf')
            
            left = helper(root.left)
            right = helper(root.right)
            ## error: just to decide to add the root or not.
            siglepath = max(left[0] + root.val, right[0] + root.val, 0)
            maxpath = max(left[1], right[1], left[0] + root.val + right[0])
            return siglepath, maxpath
        __, ret = helper(root)
        return ret
            
            
                